datalad_next/tree.py

```diff
@@ -0,0 +1,1130 @@
+                (path / ".datalad" / "metadata").is_dir():
+            # could also query `ds.id`, but checking just for existence
+            # of config file is quicker.
+            return True
+
+        # if it is not installed, check if it has an installed superdataset.
+        # instead of querying ds.is_installed() (which checks if the
+        # directory has the .git folder), we check if the directory
+        # is empty (faster) -- as e.g. after a non-recursive `datalad clone`
+        if is_empty_dir(path):
+            if get_superdataset(path) is not None:
+                return True
+
+    except Exception as ex:
+        # if anything fails (e.g. permission denied), we raise exception
+        # instead of returning False. this can be caught and handled by the
+        # caller.
+        raise NoDatasetFound(f"Cannot determine if '{path.name}' is a "
+                             f"dataset") from ex
+
+    return False
+
+
+@lru_cache
+def get_subds_paths(ds_path: Path):
+    """Return paths of immediate subdatasets for a given dataset path."""
+    # This is an expensive operation because it calls git to read the
+    # submodules. Since we need to run it to (A) calculate dataset depth and
+    # (B) detect non-installed datasets, we cache results, so that the list of
+    # subdatasets is computed only once for each parent dataset.
+
+    def res_filter(res):
+        return res.get('status') == 'ok' and res.get('type') == 'dataset'
+
+    # call subdatasets command instead of dataset method `ds.subdatasets()`
+    # to avoid potentially expensive import of full datalad API
+    return Subdatasets.__call__(
+        dataset=ds_path,
+        recursive=False,
+        state='any',  # include not-installed subdatasets
+        result_filter=res_filter,
+        on_failure='ignore',
+        result_xfm='paths',
+        result_renderer='disabled',
+        return_type='list'
+    )
+
+
+def is_subds_of_parent(subds_path: Path, parent_path: Path):
+    return any(
+        is_path_relative_to(Path(p), subds_path)
+        for p in get_subds_paths(parent_path)
+    )
+
+
+def get_dataset_root_datalad_only(path: Path):
+    """Get root of dataset containing a given path (datalad datasets only,
+    not pure git/git-annex repo)
+
+    Parameters
+    ----------
+    path: Path
+        Path to file or directory
+
+    Returns
+    -------
+    Path
+    """
+    ds_root = path
+    while ds_root:
+        potential_ds_root = get_dataset_root(str(ds_root))
+
+        if potential_ds_root is None:
+            return None  # we are not inside a dataset
+
+        potential_ds_root = Path(potential_ds_root)
+        if is_dataset(potential_ds_root):
+            return potential_ds_root  # it's a match
+
+        # we go one directory higher and try again
+        ds_root = Path.resolve(potential_ds_root / '..')
+    return ds_root
+
+
+@lru_cache
+def get_superdataset(path: Path):
```

Comment:
I think it would be worthwhile to develop this function to be a full replacement of `Dataset.get_superdataset()` and patch the latter away with the former. It would need to understand `topmost=`, `datalad|registered_only=` as additional parameters to make this happen.

This does not have to be done in this PR, but if not, should be acknowledged in a issue, such that we do not accumulate alternative definitions of what is a (super)dataset and what is not.

Reply:
https://github.com/datalad/datalad/issues/6938